File: packages/agent/agent/lib/tenant.ts
import type { SessionContext } from "eve/context";

/**
 * Identidade desta instância do agente.
 *
 * No modelo silo cada deployment atende UM tenant, e o valor vem do env do
 * projeto. Nas Etapas 0–3 (modo pool) há uma instância só, e o valor fica
 * indefinido — o guard então aceita o tenant que vier do token verificado.
 */
export function instanceTenantId(): string | undefined {
  const value = process.env.TENANT_ID;
  return typeof value === "string" && value.length > 0 ? value : undefined;
}

export type TenantCaller = { tenantId: string; userId: string };

/**
 * O mínimo que o guard precisa enxergar.
 *
 * Declarado estruturalmente porque ele roda em dois lugares com contextos
 * diferentes: nas tools (`SessionContext`) e nos resolvedores de instruções
 * dinâmicas (`DynamicResolveContext`, que não tem sandbox nem skills). Ambos
 * carregam a sessão autenticada, que é a única coisa lida aqui — exigir o
 * contexto completo obrigaria a duplicar o guard, e guard duplicado é guard
 * que uma hora diverge.
 */
export type AuthenticatedContext = {
  session: {
    id?: string;
    auth: SessionContext["session"]["auth"];
    parent?: { sessionId?: string };
  };
};

/**
 * Guard obrigatório na entrada de TODA tool.
 *
 * O tenantId nunca vem do input da tool — vem da sessão autenticada. Um
 * `tenantId` em `inputSchema` seria bug de segurança, não parâmetro.
 *
 * O `typeof === "string"` não é zelo excessivo: `attributes` é
 * `Record<string, unknown>` e os próprios exemplos da doc do eve passam arrays.
 */
export function requireTenantCaller(ctx: AuthenticatedContext): TenantCaller {
  // Declared subagents can inherit the authenticated user as the session
  // initiator while `auth.current` is reserved for the currently executing
  // identity. Treat both as the same caller for tenant-scoped operations,
  // preserving the fail-closed checks below when neither is a user.
  const caller = ctx.session.auth.current ?? ctx.session.auth.initiator;
  const tenantId = caller?.attributes?.tenantId;

  if (caller?.principalType !== "user" || typeof tenantId !== "string") {
    throw new Error("An authenticated tenant user is required.");
  }

  // Silo: a instância recusa qualquer tenant que não seja o dela, mesmo que o
  // token esteja corretamente assinado.
  const instance = instanceTenantId();
  if (instance !== undefined && tenantId !== instance) {
    throw new Error("Cross-tenant call rejected.");
  }

  return { tenantId, userId: caller.principalId };
}

export type SessionCaller = TenantCaller & { sessionId: string };

/** Uma referência conversacional só é segura quando pertence a uma sessão. */
export function requireSessionCaller(ctx: AuthenticatedContext): SessionCaller {
  const caller = requireTenantCaller(ctx);
  const sessionId = ctx.session.id;
  if (typeof sessionId !== "string" || sessionId.length === 0) {
    throw new Error("An Eve session id is required.");
  }
  return { ...caller, sessionId };
}

/** Extrai o tenantId de um principal, tolerando ausência e tipo inesperado. */
export function tenantIdOf(
  principal: SessionContext["session"]["auth"]["current"],
): string | undefined {
  const value = principal?.attributes?.tenantId;
  return typeof value === "string" ? value : undefined;
}

File: services/agent/clara/tools/tenant.py
"""Guard obrigatório na entrada de TODA tool — porta de `agent/lib/tenant.ts`.

O tenant nunca vem do input da tool: vem das `dependencies` do `RunContext`,
que o `JWTMiddleware` preenche a partir da claim verificada do token
(`dependencies_claims=["tenantId", "userId"]` em `os_app.py`). Um `tenant_id`
declarado no schema de uma tool seria bug de segurança, não parâmetro (FR-001).
"""

from __future__ import annotations

from dataclasses import dataclass

from agno.run.base import RunContext

from clara.settings import get_settings


def instance_tenant_id() -> str | None:
    """No modelo silo cada deployment atende UM tenant. Em modo pool, None —
    o guard então aceita o tenant que vier do token verificado."""
    value = get_settings().tenant_id
    return value if value else None


@dataclass(frozen=True)
class TenantCaller:
    tenant_id: str
    user_id: str


class UnauthenticatedTenantError(RuntimeError):
    pass


class CrossTenantCallError(RuntimeError):
    pass


def require_tenant_caller(run_context: RunContext) -> TenantCaller:
    dependencies = run_context.dependencies or {}
    tenant_id = dependencies.get("tenantId")
    user_id = run_context.user_id

    valid_tenant = isinstance(tenant_id, str) and tenant_id
    valid_user = isinstance(user_id, str) and user_id
    if not valid_tenant or not valid_user:
        raise UnauthenticatedTenantError("An authenticated tenant user is required.")

    # Silo: a instância recusa qualquer tenant que não seja o dela, mesmo com
    # um token corretamente assinado por outro tenant.
    instance = instance_tenant_id()
    if instance is not None and tenant_id != instance:
        raise CrossTenantCallError("Cross-tenant call rejected.")

    return TenantCaller(tenant_id=tenant_id, user_id=user_id)


@dataclass(frozen=True)
class SessionCaller(TenantCaller):
    session_id: str


def require_session_caller(run_context: RunContext) -> SessionCaller:
    """Uma referência conversacional só é segura quando pertence a uma sessão."""
    caller = require_tenant_caller(run_context)
    session_id = run_context.session_id
    if not session_id:
        raise UnauthenticatedTenantError("An agent session id is required.")
    return SessionCaller(tenant_id=caller.tenant_id, user_id=caller.user_id, session_id=session_id)
